schema: add @bracket.slope in att.tuplet.vis

pe-ro opened the pull request:
To avoid duplicate attribute definitions, att.tuplet.vis is referenced by att.tupletSpan.vis.

Fixes #1195

pe-ro committed:
Add @bracket.slope in att.tuplet.vis

To avoid duplicate attribute definitions, att.tuplet.vis is referenced by att.tupletSpan.vis.

craigsapp commented on source/modules/MEI.visual.xml:
```diff
@@ -1933,6 +1933,13 @@
       </attDef>
+      <attDef ident="bracket.slope" usage="opt">
+         <desc xml:lang="en">Records the slope of the bracket independently of the slope of the beam (if any)
+           encompassing the events of the tuplet.</desc>
```
It might be good to explicitly state the units (degrees presumably).

rettinghaus replied:
refers to #648

pe-ro replied:
Degrees sounds good. Other thoughts, @lpugin, @rettinghaus, @kepper, others?

craigsapp replied:
The problem I have in #648, is that a non-zero degree value is not overly useful since it would only apply to one specific layout (that was specifically for beams, but also applies to tuplet brackets and brackets in general).   

A better unit would be the vertical offset of the endpoints, which remains constant over all layout spacings (more or less, sine beams may be flattened when the beam angle is too high).

In other words, perhaps something better than degrees would be in diatonic steps (MEI units).   0 means horizontal, 2 = move up by a staff line from the starting point to the ending point.

pe-ro replied:
Is this description better --

Records the slope of a bracket in degrees of deviation from horizontal. Positive values indicate upward slope, negative values measure slope downward, and a zero value indicates the bracket is horizontal.  The value here is independent of the slope of the beam (if any) encompassing the events of the tuplet.

rettinghaus replied:
Degrees is problematic, because the angle depends on the spacing of the music (as seen in #648). Also you probably would want to limit it then (-90 to 90)? Slope itself is ∆y/∆x, and thus also dependent on spacing. We could define it in MEI as the difference of the distances between the endpoints relative to the staff lines. "3" would mean the end of the bracket is 3 vus higher than its start (at a horizontal staff), "-2.5" it would go down two and a half units.

pe-ro replied:
Ok,  how about modeling `@bracket.slope` on `@vo`:

```
<attDef ident="slope" usage="opt">
  <desc xml:lang="en">Records the slope of a bracket in terms of staff 
interline distance; that is, in units of 1/2 the distance between adjacent staff 
lines. Positive values indicate upward slope, negative values measure slope 
downward, and a zero value indicates the bracket is horizontal. The value here 
is independent of the slope of the beam (if any) encompassing the events of 
the tuplet.</desc>
  <datatype>
    <rng:ref name="data.MEASUREMENTSIGNED"/>
  </datatype>
</attDef>
```
To keep the description short, the text starting with "Positive values indicate ..." can go in the `<remarks>` section.

ahankinson replied:
Would `beam@slope` also change accordingly? Would be annoying to have two attributes with the same name and different types of values.

craigsapp replied:
> Ok, how about modeling @bracket.slope on `@vo`

That sounds good since a degree value will not be useful for dynamic layout control.

One thought is that `beam@slope` and similar could be changed to `@vo2` in a similar manner to `@tstamp` and `@tstamp2`:  `@vo2` would be the vertical offset of the endpoint compared to the starting point's vertical placement.   This would be clearer than using the name "slope" since that mathematically implies units of degrees.

Slur currently has `@startvo` and `@vo`, so perhaps change these to `@vo` and `@vo2`.  This would match `@tstamp` attribute naming system as well as to `@x`/`@x2` and `@y`/`@y2`.

rettinghaus replied:
@craigsapp please stop mixing things up again. `vo` is a vertical offset, accompanied by `@startvo` and `@endvo`. 
@pe-ro yes, please change beam@slope.

ahankinson commented:
Given that this is new functionality I would suggest holding off merging this until after 5.0.

rettinghaus commented:
@ahankinson actually #648 is open for over three years. It would be nice to close this now with 5.0.

ahankinson commented:
@rettinghaus an open issue doesn’t cause problems, but a hasty solution committed in a rush might, particularly if the conversation involves new ways of measurement and isn’t unified across the spec.

rettinghaus commented:
@ahankinson generally I agree with you. But the older issue pointed to the problem, which is, that the values in `@slope` didn't have a particular _meaning_, making it basically unusable. Changing it to `data.MEASUREMENTSIGNED` would only fill it with meaning without making existing encodings invalid. So this seems to me not to be a "hasty solution committed in a rush". 
But as you pointed out (and I emphasized) `@bracket.slope` and `beam@slope` need to use the same `data type`.

pe-ro commented:
I assigned this to the 5.x milestone, but I'd be happy to move it forward to 5.0.

pe-ro committed:
Add @bracket.slope and modify @beam.slope

in att.tuplet.vis
- enhance description of `@bracket.slope`
- add clarifying remark re: slope of tuplet beams

in att.beaming.vis
- harmonize description and datatype of `@beam.slope`

pe-ro committed:
Modify att.beamRend

harmonize `@slope` with new descriptions in MEI.visual.

pe-ro committed:
Fix typo

pe-ro commented:
PR updated.

craigsapp commented:
> I assigned this to the 5.x milestone, but I'd be happy to move it forward to 5.0.

I don't see a need to rush.  It only became interesting due to refinements in verovio layout (and such refinements can go on forever :-)

ahankinson commented on source/modules/MEI.cmn.xml:
```diff
@@ -291,9 +291,12 @@
-        <desc xml:lang="en">Records the slope of the beam.</desc>
+        <desc xml:lang="en">Records the slope of a beam; e.g., in terms of staff interline distance; 
+          that is, in units of 1/2 the distance between adjacent staff lines. Positive values indicate 
+          upward slope, negative values measure slope downward, and a zero value indicates the beam 
```
```suggestion
          upward slope, negative values indicate slope downward, and a zero value indicates the beam 
```

ahankinson commented on source/modules/MEI.cmn.xml:
```diff
@@ -291,9 +291,12 @@
       </attDef>
       <attDef ident="slope" usage="opt">
-        <desc xml:lang="en">Records the slope of the beam.</desc>
+        <desc xml:lang="en">Records the slope of a beam; e.g., in terms of staff interline distance; 
```
We should refer to these as virtual units (`vu`) instead of re-explaining what a VU is in MEI.

pe-ro committed:
Modify description of beam/@slope

Co-authored-by: Andrew Hankinson <andrew.hankinson@rism.digital>

pe-ro committed:
Modify description of beam/@slope

Co-authored-by: Andrew Hankinson <andrew.hankinson@rism.digital>

pe-ro committed:
Modify description of @bracket.slope

Co-authored-by: Andrew Hankinson <andrew.hankinson@rism.digital>

pe-ro committed:
Modify description of @slope

pe-ro committed:
Modify description of @bracket.slope

pe-ro commented:
I think I incorporated @ahankinson's changes, but along the way I managed to create an unnecessary branch I don't know how to remove.

ahankinson commented on source/modules/MEI.cmn.xml:
```diff
@@ -291,9 +291,11 @@
       </attDef>
       <attDef ident="slope" usage="opt">
-        <desc xml:lang="en">Records the slope of the beam.</desc>
+        <desc xml:lang="en">Records the slope of a beam; e.g., in MEI virtual units. Positive values indicate 
```
Is it really "e.g., in MEI virtual units" or just "Records the slope of a beam in MEI virtual units." ? If the latter, this should be changed in the other descriptions as well.

rettinghaus replied:
`data.MEASUREMENTSIGNED` allows all sorts of units.

ahankinson replied:
Hm. Right. OK

ahankinson reviewed (approved):
The `<abbr>` elves might make a visit, but it looks good to me!

bwbohl commented:
Generally I do agree but would love to see some examples and documentation added to the guidelines ;-)

musicEnfanthen commented:
Still needs work, moving to 6.

bwbohl commented:
I think adding some minimal mention of bracket.slope to the guidelines would do the job of a minimum documentation

bwbohl committed:
Merge branch 'develop' into develop-tuplet-slope

bwbohl commented:
@pe-ro, am I assuming correctly that in the context of this pull request, the following branch can be removed: https://github.com/music-encoding/music-encoding/tree/pe-ro-patch-1?

pe-ro commented:
@bwbohl , you're correct

bwbohl committed:
Merge branch 'develop' into develop-tuplet-slope
